corehq/apps/users/views/mobile/users.py

```diff
@@ -826,6 +727,88 @@ def _construct_form_data(self, in_data, fields):
+
+
+@require_can_edit_commcare_users
+@require_POST
+def deactivate_commcare_user(request, domain, user_id):
+    return _modify_user_status(request, domain, user_id, False)
+
+
+def _modify_user_status(request, domain, user_id, is_active):
+    user = CommCareUser.get_by_user_id(user_id, domain)
+    if (not _can_edit_workers_location(request.couch_user, user)
+            or (is_active and not can_add_extra_mobile_workers(request))):
+        return json_response({
+            'error': _("No Permission."),
+        })
+    if not is_active and user.user_location_id:
+        return json_response({
+            'error': _("This is a location user, archive or delete the "
+                       "corresponding location to deactivate it."),
+        })
+    user.is_active = is_active
+    user.save()
+    return json_response({
+        'success': True,
+    })
+
+
+@require_can_edit_commcare_users
+@require_GET
+def paginate_mobile_workers(request, domain):
+    limit = int(request.GET.get('limit', 10))
+    page = int(request.GET.get('page', 1))
+    query = request.GET.get('query')
+    deactivated_only = bool(int(request.GET.get('showDeactivatedUsers', 0)))
+
+    def _user_query(search_string, page, limit):
+        user_es = get_search_users_in_domain_es_query(
+            domain=domain, search_string=search_string,
+            offset=page * limit, limit=limit)
+        if not request.couch_user.has_permission(domain, 'access_all_locations'):
+            loc_ids = (SQLLocation.objects.accessible_to_user(domain, request.couch_user)
+                                          .location_ids())
+            user_es = user_es.location(list(loc_ids))
+        return user_es.mobile_users()
+
+    def _format_user(user_json):
+        user = CouchUser.wrap_correctly(user_json)
+        return {
+            'username': user.raw_username,
+            'first_name': user.first_name,
+            'last_name': user.last_name,
+            'phoneNumbers': user.phone_numbers,
+            'user_id': user.user_id,
+            'mark_activated': False,
+            'mark_deactivated': False,
+            'dateRegistered': user.created_on.strftime(USER_DATE_FORMAT) if user.created_on else '',
+            'editUrl': reverse(EditCommCareUserView.urlname, args=[domain, user.user_id]),
+            'deactivateUrl': "#",
+            'actionText': _("Deactivate") if user.is_active else _("Activate"),
+            'action': 'deactivate' if user.is_active else 'activate',
+        }
+
+    # backend pages start at 0
+    users_query = _user_query(query, page - 1, limit)
+    # run with a blank query to fetch total records with same scope as in search
+    total_records = _user_query('', 0, 0).count()
+    if deactivated_only:
+        users_query = users_query.show_only_inactive()
+    users_data = users_query.run()
+    return json_response({
+        'users': [_format_user(user) for user in users_data.hits],
```

Comment:
so I think, since you are only using a few fields from the user es model, you could actually do: 
```py
users_data = users_query.source(['first_name', 'last_name', 'raw_username', '_id', 'created_on']).run()
``` 
and it will pre-format it for you and only pull those values from ES (i.e. it needs to pull less data from ES)

Reply:
Because `_format_user` calls `wrap`, I need to either stop wrapping (and duplicate `raw_username` logic) or include the various fields needed to wrap, and both of those options are messier than the current code.

Reply:
`base_username` in elasticsearch is `raw_username` on the wrapped document (https://github.com/dimagi/commcare-hq/blob/465c6380133d291df51b9433da90ea9d4c40a832/corehq/pillows/user.py#L49-L52) Why? Who knows.
I'm saying  you can just remove the whole `_format_user` function if you pull out just the fields that you want from ES, and use its output in your view. This means you don't have to wrap each user, and you only retrieve the amount of data you actually are using from ES. 
```py
In [14]: es.UserES().mobile_users().fields(['first_name', 'last_name', 'base_username', '_id', 'created_on', 'is_active']).run().hits[5]
Out[14]: 
{u'_id': u'1ef76033c5a64b659fac741ac4fb2c62',
 u'base_username': u'test2',
 u'created_on': u'2018-10-03T18:18:59.194435Z',
 u'first_name': u'',
 u'is_active': True,
 u'last_name': u''}
```

Reply:
Aha, knowing about `base_username` was the missing link. 91ee416